Add option to specify if the customer choose to receive available products

thalassa-web opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | In PrestaShop 1.6 the customer was able to choose to recieve available products first. The option is still available in cart but the controller do not use it anymore.
| Type?             | bug fix
| Category?         |  CO
| BC breaks?        | no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #9781
| How to test?      | Add an available product and an unavailable product (but orderable) in the cart. Enter in the checkout tunnel. In the shipping step, check the checkbox "Send available products first". The order should be split in two.
| Possible impacts? | The option is inactive by default, impacts are very limited.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/22941)
<!-- Reviewable:end -->

thalassa-web committed:
Add option to specify if the customer choose te receive available products before unavailable ones.
This option update cart and specify the allow_seperated_package property.
Add the checkbox in the classic theme's shipping step.
I do not know how to add the French translation …

matks commented on themes/classic/templates/checkout/_partials/steps/shipping.tpl:
```diff
@@ -105,6 +105,15 @@
 
+            {if $separatePackages.allowed}
+              <span class="custom-checkbox">
+                <input class="js-gift-checkbox" id="allow_seperated_package" name="allow_seperated_package"
```
```suggestion
                <input class="js-gift-checkbox" id="allow_separated_package" name="allow_separated_package"
```

matks commented on themes/classic/templates/checkout/_partials/steps/shipping.tpl:
```diff
@@ -105,6 +105,15 @@
+                <input class="js-gift-checkbox" id="allow_seperated_package" name="allow_seperated_package"
+                       type="checkbox" value="1" {if $separatePackages.isSeparatePackage}checked="checked"{/if}>
+                <span><i class="material-icons rtl-no-flip checkbox-checked">&#xE5CA;</i></span>
+                <label for="allow_seperated_package">{l s='Send available products first' d='Shop.Theme.Checkout'}</label>
```
```suggestion
                <label for="allow_separated_package">{l s='Send available products first' d='Shop.Theme.Checkout'}</label>
```

Julievrz replied:
```suggestion
                <label for="allow_seperated_package">{l s='Receive available products first' d='Shop.Theme.Checkout'}</label>
```

matks commented on classes/checkout/CheckoutDeliveryStep.php:
```diff
@@ -136,6 +149,9 @@ public function handleRequest(array $requestParams = [])
                 ($useGift && isset($requestParams['gift_message'])) ? $requestParams['gift_message'] : ''
             );
+            $this->getCheckoutSession()->setSeparatePackagesAllowed(
+                isset($requestParams['allow_seperated_package']) ? $requestParams['allow_seperated_package'] : false
```
```suggestion
                isset($requestParams['allow_separated_package']) ? $requestParams['allow_separated_package'] : false
```
Little typo 😉

Progi1984 replied:
```suggestion
                $requestParams['allow_separated_package']) ?? false
```

matks commented on classes/checkout/CheckoutDeliveryStep.php:
```diff
@@ -120,6 +121,18 @@ public function getGiftCostForLabel()
         return '';
     }
 
+    public function setSeparatePackagesAllowed($separatePackagesAllowed)
```
Hi, I think it could be useful to add phpDoc 😊  ?

I don't know if we can type arguments and return type 🤔  wdyt @Progi1984 ? should we introduce strong type hint even if rest of the class does not use it ?

Progi1984 replied:
```suggestion
    public function setSeparatePackagesAllowed(bool $separatePackagesAllowed): self
```

matks commented on classes/checkout/CheckoutSession.php:
```diff
@@ -196,4 +196,15 @@ public function getCheckoutURL()
+
+    public function setSeparatePackagesAllowed($separatePackagesAllowed)
+    {
+        $this->context->cart->allow_seperated_package = $separatePackagesAllowed;
```
```suggestion
        $this->context->cart->allow_seperated_package = (bool) $separatePackagesAllowed;
```
If we do not force the type to bool, maybe we should cast it here to make sure we don't mess the cart 👀

matks replied:
Also typo again 😅 
```suggestion
        $this->context->cart->allow_separated_package = $separatePackagesAllowed;
```

matks reviewed (commented):
Hi @thalassa-web thank you for the PR it is a lot better now 🎉 

I added a few suggestions to make the code even better.

So about the behavior: this PR makes the option
- visible in checkout (customer can decide to enable it or not)
- you say the backend logic is active so if option is enabled by customer, the order will be automatically split when confirmed?
- maybe we should also make sure this is correctly being displayed in back-office? merchant should see that the order has been split to allow available products being shipped first?

ping @PrestaShop/product-team who might want to discuss about the behavior :)

LouiseBonnard commented:
@matks, @thalassa-web, following the current split order behavior in the BO, the two orders should share the name reference.

Progi1984 commented on classes/checkout/CheckoutDeliveryStep.php:
```diff
@@ -34,6 +34,7 @@ class CheckoutDeliveryStepCore extends AbstractCheckoutStep
     private $giftCost = 0;
     private $includeTaxes = false;
     private $displayTaxesLabel = false;
+    private $separatePackagesAllowed = false;
```
```suggestion
    /**
     * @var bool
     */
    private $separatePackagesAllowed = false;
```

Progi1984 commented on classes/checkout/CheckoutDeliveryStep.php:
```diff
@@ -120,6 +121,18 @@ public function getGiftCostForLabel()
+        return $this;
+    }
+
+    public function isSeparatePackagesAllowed()
```
```suggestion
    public function isSeparatePackagesAllowed(): bool
```

Progi1984 commented on classes/checkout/CheckoutSession.php:
```diff
@@ -196,4 +196,15 @@ public function getCheckoutURL()
+        return $this->context->cart->update();
+    }
+
+    public function isSeparatePackagesAllowed()
```
```suggestion
    public function isSeparatePackagesAllowed(): bool
```

Progi1984 reviewed (commented):
Some feedbacls

Julievrz commented:
> This pull request seems to contain new translation strings. I have summarized them below to ease up review:
> 
> * `Shop.Theme.Checkout`
>   
>   * [ ]  [`Send available products first`](https://github.com/PrestaShop/PrestaShop/pull/22941/files#diff-ce8fe45c8932c4dfa982ef70fb842fbdR113)
> 
> (Note: this is an automated message, but answering it will reach a real human)

As the wording will appear in the front office, doesn't it make more sense to write "Receive available products first" in Shop.Theme.Checkout?

Hlavtox commented:
@thalassa-web @matks This is an amazing feature! 🚀

MatShir commented:
@thalassa-web very cool feature, I just add waiting for UX to be sure that the location of the feature matches well with our checkout, but you still can continue on your work 🚀

prestascott commented:
Hello everyone, 

Thanks @thalassa-web for your PR! It's a great improvement. :)
Placement of the option to receive available products first is fine in the Shipping method step. I just recommend to put the option just before the comment section because it's more important, it needs to be placed higher in the information hierarchy. Comment section is used to allow customers to write additional information about the shipping, and split order can be a topic where he can write more about. 

Thanks!

marionf commented:
Hello @thalassa-web 

It sounds very good to restore the checkbox "Send available products first" at the shipping step during the checkout process :+1: 
Could you handle [the feedback](https://github.com/PrestaShop/PrestaShop/pull/22941#issuecomment-774138671)  of @prestascott and the code & wording changes requested ?

kpodemski commented:
Hello @thalassa-web 

Good work, do you plan to address feedback from our team? It would be great to have this feature back

matks commented:
Ping @thalassa-web friendly reminder 😉

PierreRambaud commented:
> Hi everyone …
> Sorry to respond just now … did'nt receive any e-mail about comments on this PR … (I should check my email config …)
> Just to be sure, because none of my previous PR was sooo followed and with some changes, I need some help to know what I have to do with this PR …
> Specificly with my IDE and git …

You can peek proposed changes, add them into your branch, and push it :)

thalassa-web committed:
Update themes/classic/templates/checkout/_partials/steps/shipping.tpl

Suggested chgange by matks

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

thalassa-web committed:
Update themes/classic/templates/checkout/_partials/steps/shipping.tpl

Suggestion by matks

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

thalassa-web committed:
Update classes/checkout/CheckoutDeliveryStep.php

Suggestion by matks

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

thalassa-web committed:
Update classes/checkout/CheckoutSession.php

Suggestion by matks

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

thalassa-web committed:
Update classes/checkout/CheckoutDeliveryStep.php

Suggestion by Progi1984

Co-authored-by: Progi1984 <progi1984@gmail.com>

thalassa-web committed:
Update classes/checkout/CheckoutDeliveryStep.php

Suggestion by Progi1984

Co-authored-by: Progi1984 <progi1984@gmail.com>

thalassa-web committed:
Update classes/checkout/CheckoutDeliveryStep.php

Suggestion by Progi1984

Co-authored-by: Progi1984 <progi1984@gmail.com>

thalassa-web committed:
Update classes/checkout/CheckoutSession.php

Suggestion by Prosi1984

Co-authored-by: Progi1984 <progi1984@gmail.com>

thalassa-web committed:
Manual merge of specific suggestions

thalassa-web commented:
Hi guys !
Ok it seems I found how to include suggestions in the PR (thanks to PierreRambaud).
One last question though, the bot suggested there is a new translation string (and indeed, there is).
Where can I put the new string «Receive available products first» for Shop.Theme.Checkout.
(I could only translate it in French ^-^)

thalassa-web committed:
Some coding standard fixes

Julievrz reviewed (dismissed):
Wording ✅

thalassa-web committed:
The prestascott's suggestion is quite true.
The option coul'd be more important than comments …
But of course, it's just the default template …

PierreRambaud commented on classes/checkout/CheckoutDeliveryStep.php:
```diff
@@ -136,6 +152,9 @@ public function handleRequest(array $requestParams = [])
                 ($useGift && isset($requestParams['gift_message'])) ? $requestParams['gift_message'] : ''
             );
+            $this->getCheckoutSession()->setSeparatePackagesAllowed(
+                $requestParams['allow_separated_package'] ??  false
```
```suggestion
                $requestParams['allow_separated_package'] ?? false
```

PierreRambaud commented on classes/checkout/CheckoutSession.php:
```diff
@@ -196,4 +196,16 @@ public function getCheckoutURL()
         return $this->context->link->getPageLink('order');
     }
+
+    public function setSeparatePackagesAllowed($separatePackagesAllowed)
```
```suggestion
    public function setSeparatePackagesAllowed(bool $separatePackagesAllowed): bool
```
?

PierreRambaud commented on classes/checkout/CheckoutSession.php:
```diff
@@ -196,4 +196,16 @@ public function getCheckoutURL()
+
+    public function setSeparatePackagesAllowed($separatePackagesAllowed)
+    {
+        $this->context->cart->allow_separated_package = (bool) $separatePackagesAllowed;
```
```suggestion
        $this->context->cart->allow_separated_package = $separatePackagesAllowed;
```

thalassa-web committed:
Hard type hitting

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

thalassa-web committed:
Too much space

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

thalassa-web committed:
Attribute is obviously bool

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

thalassa-web commented:
@matks waiting for some news about this PR …

kpodemski commented:
Hello @thalassa-web 

Could you fix the issues coming from cs-fixer? :-)

kpodemski commented on classes/checkout/CheckoutDeliveryStep.php:
```diff
@@ -187,6 +206,10 @@ public function render(array $extraParams = [])
                 ],
+                'separatePackages' => [
+                    'allowed' => $this->isSeparatePackagesAllowed(),
+                    'isSeparatePackage' => $this->getCheckoutSession()->isSeparatePackagesAllowed()
```
```suggestion
                    'isSeparatePackage' => $this->getCheckoutSession()->isSeparatePackagesAllowed(),
```

thalassa-web committed:
Update classes/checkout/CheckoutDeliveryStep.php

CS-fixer fix …

Co-authored-by: Krystian Podemski <kpodemski@users.noreply.github.com>

thalassa-web commented:
It's quite annoying … I haven't made any changing about wording with the last commit, but I have to ask the approval again …

kpodemski commented on themes/classic/templates/checkout/_partials/steps/shipping.tpl:
```diff
@@ -79,6 +79,15 @@
           <div class="order-options">
+            {if $separatePackages.allowed}
+              <span class="custom-checkbox">
+              <input class="js-gift-checkbox" id="allow_separated_package" name="allow_separated_package"
```
@thalassa-web can we change this class? I think it should be related to the separated package checkbox

thalassa-web replied:
After some discussion with myself and I … You're right !

kpodemski commented on themes/classic/templates/checkout/_partials/steps/shipping.tpl:
```diff
@@ -79,6 +79,15 @@
           <div class="order-options">
+            {if $separatePackages.allowed}
+              <span class="custom-checkbox">
+              <input class="js-gift-checkbox" id="allow_separated_package" name="allow_separated_package"
```
```suggestion
              <input id="allow_separated_package" name="allow_separated_package"
```

In the end, I don't we need it

kpodemski replied:
but don't trust me with this one, if there are some visual aspects related to this class, we need to make sure that we're add a dedicated one

thalassa-web replied:
Honestly, I think it comes from the source code I copied from PrestaShop 1.6.
This PR has almost 1 year now … and some reviewed was made.
If the class is not used, I know it makes some dead code … but It worked as this … Now, I have to check if the removal of the class is ok or not. It will dellay (again) the resolution for just a class attribute …

thalassa-web replied:
There is some js code which use this class : 
themes/classic/_dev/js/checkout.js
themes/classic/assets/js/theme.js
tests/Selenium/specs/checkout.js

So I don't think we have to remove the class.

thalassa-web replied:
Or maybe I missused the class …
Is it the class for the gift checkbox maybe ?
Copy/Paste error ?

thalassa-web replied:
I think you're right … We should remove this class …

Julievrz commented:
> It's quite annoying … I haven't made any changing about wording with the last commit, but I have to ask the approval again …

I know :-( The bot is not always right. (Maybe @atomiix has an idea?) I approved the wording of this PR again.

kpodemski commented:
@Julievrz, I think it would be pretty complicated to handle this, but maybe we can agree that one of the maintainers can set this label to `wording ok` if nothing changes?

thalassa-web committed:
Update themes/classic/templates/checkout/_partials/steps/shipping.tpl

Copy/Paste problem … This class is used for the gift checkbox and does not related to the separate_package one …

Co-authored-by: Krystian Podemski <kpodemski@users.noreply.github.com>

thalassa-web reviewed (commented):
I do not know yet … What is the point with the Review …

kpodemski commented:
@thalassa-web I know that it is a bit annoying that it all takes so much time. Thank you very much for your patience and contribution 💪

kpodemski commented:
@thalassa-web, I don't think your change is enough to make this mechanism work after further checking this topic. There's much more to it. I'm afraid that it would be hard to make it work with 1.7 without additional work :-(

thalassa-web commented:
@kpodemski I tried 1 year ago on a 1.7.x and it worked well because lot of work was already make since the 1.6 version …
I tried on my PerstaShop and the order was splitted …
Can you give me some clues about why it won't work ?

kpodemski commented:
@thalassa-web I just double-checked `Cart` class, and there's indeed an `allow_seperated_package`, I must have searched for the wrong phrase. Sorry for the confusion

atomiix commented:
@thalassa-web @Julievrz the label "Waiting for wording" is added each time a PR contains a wording (new or old) and each time a new commit is made as it can change the context of the wording and may require a new validation. The post made by prestonbot with checkboxes though, should keep track of validated new wording even when new commits are made. From what I see here, @Julievrz checked the checkbox **at the same time** prestonbot updated the comment, that's probably why it didn't stay checked.

PierreRambaud commented on themes/classic/templates/checkout/_partials/steps/shipping.tpl:
```diff
@@ -79,6 +79,15 @@
             </div>
           {/block}
           <div class="order-options">
+            {if $separatePackages.allowed}
+              <span class="custom-checkbox">
+              <input id="allow_separated_package" name="allow_separated_package"
+                     type="checkbox" value="1" {if $separatePackages.isSeparatePackage}checked="checked"{/if}>
+              <span><i class="material-icons rtl-no-flip checkbox-checked">&#xE5CA;</i></span>
+              <label for="allow_separated_package">{l s='Receive available products first' d='Shop.Theme.Checkout'}</label>
+            </span>
+            {/if}
```
```suggestion
            {if $separatePackages.allowed}
              <span class="custom-checkbox">
                <input id="allow_separated_package" name="allow_separated_package"
                  type="checkbox" value="1" {if $separatePackages.isSeparatePackage}checked="checked"{/if}>
                <span><i class="material-icons rtl-no-flip checkbox-checked">&#xE5CA;</i></span>
                <label for="allow_separated_package">{l s='Receive available products first' d='Shop.Theme.Checkout'}</label>
              </span>
            {/if}
```

PierreRambaud replied:
Indentation seems broken :(

thalassa-web committed:
Update themes/classic/templates/checkout/_partials/steps/shipping.tpl

Indentation fix

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

Progi1984 commented on classes/checkout/CheckoutSession.php:
```diff
@@ -196,4 +196,16 @@ public function getCheckoutURL()
+
+    public function isSeparatePackagesAllowed(): bool
+    {
+        return $this->context->cart->allow_seperated_package;
```
```suggestion
        return $this->context->cart->allow_separated_package;
```

thalassa-web replied:
Why not but in this case we have to change the property in the Cart class …
Or I didn't change this !!!
If I commit this, the fonctionnality will break …

thalassa-web replied:
It's an existing typo error in the PrestaShop project …
Check on the main branch ;)

Progi1984 replied:
Ok. In this case the l. 202 is not good ;)

Progi1984 commented on themes/classic/templates/checkout/_partials/steps/shipping.tpl:
```diff
@@ -79,6 +79,15 @@
+            {if $separatePackages.allowed}
+              <span class="custom-checkbox">
+                <input id="allow_separated_package" name="allow_separated_package"
+                  type="checkbox" value="1" {if $separatePackages.isSeparatePackage}checked="checked"{/if}>
```
```suggestion
                  type="checkbox" value="1"{if $separatePackages.isSeparatePackage} checked="checked"{/if} />
```

thalassa-web replied:
I am not against to fix errors, but I begin to think you do not want of this PR.
Indeed, I completed the tpl file with some copy/paste of the original code …
In the same file, we have inputs radio with this line of code :
```html
<input type="radio" name="delivery_option[{$id_address}]" id="delivery_option_{$carrier.id}"
                               value="{$carrier_id}"{if $delivery_option == $carrier_id} checked{/if}>
```
I think it's good to respect xhtml but it is also good to have consistency in rules …

Progi1984 replied:
It was a bonus not a Must Have ;)

Progi1984 reviewed (commented):
@thalassa-web Thank you for your contribution. I have detected a small bug. Could you fix it ?

Progi1984 commented on classes/checkout/CheckoutSession.php:
```diff
@@ -196,4 +196,16 @@ public function getCheckoutURL()
+
+    public function setSeparatePackagesAllowed(bool $separatePackagesAllowed): bool
+    {
+        $this->context->cart->allow_separated_package = $separatePackagesAllowed;
```
@thalassa-web This one is the bad property :

```suggestion
        $this->context->cart->allow_seperated_package = $separatePackagesAllowed;
```

thalassa-web replied:
You are right … I missread this …
I thought I used the good property but I did not …
Thanks

thalassa-web committed:
Update classes/checkout/CheckoutSession.php

Typo error

Co-authored-by: Progi1984 <progi1984@gmail.com>

thalassa-web reviewed (commented):
It seems review is ok

marwachelly commented:
Hello,
I tested the following case and I noticed that the feature of splitting order is not implemented:
> Go to FO choose a customized product
> Create an order with the same customized product with different customization and different quantity (set the quantity desired in the shopping cart)
Ps: the total sum of the quantity exceeds the stock
> Check the checkbox Receive available products first
> Go to BO > order page > The order is not splitted 
![image](https://user-images.githubusercontent.com/89604350/140082797-8e24f24d-7fc4-473b-b971-cc59e92e025c.png)

https://www.awesomescreenshot.com/video/5881863?key=b72664897be5f691044bb94f6dff9532

Could you please check.
Thanks,

matks commented:
Hi,

Since we had no news from you for more than 30 days, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. Feel free to reopen or open another one if you think it's still relevant 😉 . You can ping me!

Thanks! and maybe see you soon
